helix-term/src/commands.rs

```diff
@@ -4901,6 +4905,168 @@ fn add_newline_impl(cx: &mut Context, open: Open) {
+            MoveSelection::Below => start_line_pos + cursor.col + next_line_len + ending_len,
+        }
+    }
+}
+
+/// Move line or block of text in specified direction.
+/// The function respects single line, single selection, multiple lines using
+/// several cursors and multiple selections.
+fn move_selection(cx: &mut Context, direction: MoveSelection) {
+    let (view, doc) = current!(cx.editor);
+    let selection = doc.selection(view.id);
+    let text = doc.text();
+    let slice = text.slice(..);
+    let all_changes = selection.into_iter().map(|range| {
+        let (start, end) = range.line_range(slice);
+        let line_start = text.line_to_char(start);
+        let line_end = line_end_char_index(&slice, end);
+        let line = text.slice(line_start..line_end).to_string();
+
+        let next_line = match direction {
+            MoveSelection::Above => start.saturating_sub(1),
+            MoveSelection::Below => end + 1,
+        };
+
+        if next_line == start || next_line >= text.len_lines() {
+            vec![(line_start, line_end, Some(line.into()))]
+        } else {
+            let next_line_start = text.line_to_char(next_line);
+            let next_line_end = line_end_char_index(&slice, next_line);
+
+            let next_line_text = text.slice(next_line_start..next_line_end).to_string();
+
+            match direction {
+                MoveSelection::Above => vec![
+                    (next_line_start, next_line_end, Some(line.into())),
+                    (line_start, line_end, Some(next_line_text.into())),
+                ],
+                MoveSelection::Below => vec![
+                    (line_start, line_end, Some(next_line_text.into())),
+                    (next_line_start, next_line_end, Some(line.into())),
+                ],
+            }
+        }
+    });
+
+    // Conflicts might arise when two cursors are pointing to adjacent lines.
+    // The resulting change vector would contain two changes referring the same lines,
+    // which would make the transaction to panic.
+    // Conflicts are resolved by picking only the top change in such case.
+    fn remove_conflicts(changes: Vec<Change>) -> Vec<Change> {
+        if changes.len() > 2 {
+            changes
+                .into_iter()
+                .fold_while(vec![], |mut acc: Vec<Change>, change| {
+                    if let Some(last_change) = acc.pop() {
+                        if last_change.0 >= change.0 || last_change.1 >= change.1 {
+                            acc.push(last_change);
+                            Done(acc)
+                        } else {
+                            acc.push(last_change);
+                            acc.push(change);
+                            Continue(acc)
+                        }
+                    } else {
+                        acc.push(change);
+                        Continue(acc)
+                    }
+                })
+                .into_inner()
```

Comment:
This can be written more cleanly without itertools; no need to pull in an extra dependency.

```rust
let mut new_changes: Vec<Change> = Vec::new();

for change in changes {
    match new_changes.last() {
        Some(last_change) if last_change.0 >= change.0 || last_change.1 >= change.1 => {
            return new_changes;
        }
        _ => new_changes.push(change.clone()),
    }
}

new_changes
```

Reply:
Thanks for the suggestions!

> Moving the full line when only a subset of the line is selected though conflicts with the select-then-act paradigm.

Yeah, I also wondered about this being a bit inconsistent in the sense that modifications are usually applied to the selected area. The way I solved the problem is similar to what other editors do, such as sublime text. 

If we have in mind that helix always has some selection (cursor itself), then moving not fully selected line would be conflicting with select-then-act paradigm. Do you have alternative suggestion that would keep the feature usable? 

> multi-cursor behavior (for example: C and then C-j moves the cursors oddly; C and then C-k removes one of the cursors).

As I mentioned in the issue comment, I've left some edge case unresolved until I get signal that community want feature like this to be supported at all. If we want to have full multi cursor support, I can work on these cases.

> So unless this adds anything extra like handling indentation, I don't see a need to add extra commands for this.

Again, I communicated I would like to add indentation as the next step, but there's nothing preventing me from doing that in this PR (since it's just one function).

Reply:
> This can be written more cleanly without itertools; no need to pull in an extra dependency.
> 
> ```rust
> let mut new_changes: Vec<Change> = Vec::new();
> 
> for change in changes {
>     match new_changes.last() {
>         Some(last_change) if last_change.0 >= change.0 || last_change.1 >= change.1 => {
>             return new_changes;
>         }
>         _ => new_changes.push(change.clone()),
>     }
> }
> 
> new_changes
> ```

Thanks, I'll do it!